src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryNullableDirective/CSharpRemoveUnnecessaryNullableDirectiveCodeFixProvider.cs

```diff
@@ -53,10 +57,46 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             CodeActionOptionsProvider fallbackOptions,
             CancellationToken cancellationToken)
         {
-            foreach (var diagnostic in diagnostics)
+            var nullableDirectivesByNodes = diagnostics.GroupBy(x =>
+                x.Location.FindNode(findInsideTrivia: false, getInnermostNodeForTie: false, cancellationToken), x =>
+                x.Location.FindNode(findInsideTrivia: true, getInnermostNodeForTie: true, cancellationToken))
+                .ToDictionary(x => x.Key, x => x.ToList());
+
+            foreach (var (node, nullableDirectives) in nullableDirectivesByNodes)
             {
-                var nullableDirective = diagnostic.Location.FindNode(findInsideTrivia: true, getInnermostNodeForTie: true, cancellationToken);
-                editor.RemoveNode(nullableDirective, SyntaxRemoveOptions.KeepNoTrivia);
+                if (node is CompilationUnitSyntax)
+                {
+                    foreach (var nullableDirective in nullableDirectives)
+                        editor.RemoveNode(nullableDirective, SyntaxRemoveOptions.KeepNoTrivia);
+                    continue;
+                }
+
+                var leadingTrivia = node.GetLeadingTrivia();
+                var indexes = nullableDirectives.Select(x => leadingTrivia.IndexOf(x.ParentTrivia))
+                                                .OrderByDescending(x => x).ToArray();
+                foreach (var index in indexes)
```

Comment:
i really don't understand the logic here.  including what the above indices are representing, or waht the loop below is doing.  this all needs examples.

Reply:
Added comments, does it help?